cpp/src/merge/merge.cu

```diff
@@ -0,0 +1,348 @@
+    CUDF_EXPECTS(asc_desc.size() > 0, "Empty asc_desc");
+    CUDF_EXPECTS(asc_desc.size() <= static_cast<size_t>(left_table.num_columns()), "Too many values in asc_desc");
+    CUDF_EXPECTS(key_cols.size() == asc_desc.size(), "Mismatched size between key_cols and asc_desc");
+
+
+    auto gdf_col_deleter = [](gdf_column *col) {
+        gdf_column_free(col);
+        delete col;
+    };
+    using gdf_col_ptr = typename std::unique_ptr<gdf_column, decltype(gdf_col_deleter)>;
+    std::vector<gdf_col_ptr> temp_columns_to_free;
+    std::vector<gdf_column*> left_cols_sync(const_cast<gdf_column**>(left_table.begin()), const_cast<gdf_column**>(left_table.end()));
+    std::vector<gdf_column*> right_cols_sync(const_cast<gdf_column**>(right_table.begin()), const_cast<gdf_column**>(right_table.end()));
+    for (gdf_size_type i = 0; i < left_table.num_columns(); i++) {
+        gdf_column * left_col = const_cast<gdf_column*>(left_table.get_column(i));
+        gdf_column * right_col = const_cast<gdf_column*>(right_table.get_column(i));
+        
+        if (left_col->dtype != GDF_STRING_CATEGORY){
+            continue;
+        }
+
+        // If the inputs are nvcategory we need to make the dictionaries comparable
+
+        temp_columns_to_free.push_back(gdf_col_ptr(new gdf_column{}, gdf_col_deleter));
+        gdf_column * new_left_column_ptr = temp_columns_to_free.back().get();
+        temp_columns_to_free.push_back(gdf_col_ptr(new gdf_column{}, gdf_col_deleter));
+        gdf_column * new_right_column_ptr = temp_columns_to_free.back().get();
+
+        *new_left_column_ptr = allocate_like(*left_col, true, stream);
+        if (new_left_column_ptr->valid) {
+            CUDA_TRY( cudaMemcpyAsync(new_left_column_ptr->valid, left_col->valid, sizeof(gdf_valid_type)*gdf_num_bitmask_elements(left_col->size), cudaMemcpyDefault, stream) );
+            new_left_column_ptr->null_count = left_col->null_count;
+        }
+        
+        *new_right_column_ptr = allocate_like(*right_col, true, stream);
+        if (new_right_column_ptr->valid) {
+            CUDA_TRY( cudaMemcpyAsync(new_right_column_ptr->valid, right_col->valid, sizeof(gdf_valid_type)*gdf_num_bitmask_elements(right_col->size), cudaMemcpyDefault, stream) );
+            new_right_column_ptr->null_count = right_col->null_count;
+        }
+
+        gdf_column * tmp_arr_input[2] = {left_col, right_col};
+        gdf_column * tmp_arr_output[2] = {new_left_column_ptr, new_right_column_ptr};
+        CUDF_TRY( sync_column_categories(tmp_arr_input, tmp_arr_output, 2) );
+
+        left_cols_sync[i] = new_left_column_ptr;
+        right_cols_sync[i] = new_right_column_ptr;
+    }
+
+    table left_sync_table(left_cols_sync.data(), left_cols_sync.size());
+    table right_sync_table(right_cols_sync.data(), right_cols_sync.size());
+
+    std::vector<gdf_column*> left_key_cols_vect(key_cols.size());
```

Comment:
It looks like the only reason you're creating `left_key_cols_vect` and `right_key_cols_vect` is to pass into `device_table::create`, but that's unneccessary, because you can pass the `table` directly into `device_table::create`. 

https://github.com/rapidsai/cudf/blob/branch-0.9/cpp/src/table/device_table.cuh#L94

Reply:
I need them because I'm using just the `key_cols` columns of the `left_sync_table` and `right_sync_table` not the whole table